src/charm.py

```diff
@@ -1,104 +1,123 @@
 # Copyright 2021 Canonical Ltd.
 # See LICENSE file for licensing details.
-#
-# Learn more at: https://juju.is/docs/sdk
 
-"""Charm the service.
-
-Refer to the following post for a quick-start guide that will help you
-develop a new k8s charm using the Operator Framework:
-
-    https://discourse.charmhub.io/t/4208
-"""
+"""A Juju charm for MySQL InnoDB Cluster."""
 
 import logging
+import secrets
+import string
 
-from ops.charm import CharmBase
-from ops.framework import StoredState
+from ops.charm import CharmBase, WorkloadEvent
 from ops.main import main
-from ops.model import ActiveStatus
+from ops.model import ActiveStatus, WaitingStatus
+from ops.pebble import Layer
 
 logger = logging.getLogger(__name__)
 
 
-class OperatorTemplateCharm(CharmBase):
-    """Charm the service."""
-
-    _stored = StoredState()
+class MysqlOperatorCharm(CharmBase):
+    """A Juju Charm for MySQL InnoDB Cluster."""
 
     def __init__(self, *args):
         super().__init__(*args)
-        self.framework.observe(self.on.httpbin_pebble_ready, self._on_httpbin_pebble_ready)
-        self.framework.observe(self.on.config_changed, self._on_config_changed)
-        self.framework.observe(self.on.fortune_action, self._on_fortune_action)
-        self._stored.set_default(things=[])
 
-    def _on_httpbin_pebble_ready(self, event):
-        """Define and start a workload using the Pebble API.
+        self._config = self.model.config
+
+        self.framework.observe(self.on.start, self._on_start)
+        self.framework.observe(self.on.config_changed, self._on_config_changed)
+        change_events = [
+            self.on.mysql_server_pebble_ready,
+            self.on.mysql_router_pebble_ready,
+        ]
+        for event in change_events:
+            self.framework.observe(event, self._on_pebble_ready)
+
+    # ---- Event handlers
+    def _on_start(self, _):
+        """Placeholder for StartEvent."""
+        pass
 
-        TEMPLATE-TODO: change this example to suit your needs.
-        You'll need to specify the right entrypoint and environment
-        configuration for your specific workload. Tip: you can see the
-        standard entrypoint of an existing container using docker inspect
+    def _on_config_changed(self, _):
+        """Placeholder for ConfigChangedEvent."""
+        pass
 
-        Learn more about Pebble layers at https://github.com/canonical/pebble
-        """
-        # Get a reference the container attribute on the PebbleReadyEvent
+    def _on_pebble_ready(self, event: WorkloadEvent) -> None:
+        """Updates the Pebble configuration layer when changed."""
         container = event.workload
-        # Define an initial Pebble layer configuration
-        pebble_layer = {
-            "summary": "httpbin layer",
-            "description": "pebble config layer for httpbin",
+        if not container.can_connect():
+            self.unit.status = WaitingStatus("Waiting for pod startup to complete")
+            return
```

Comment:
Am I wrong if I say that this return should be changed to a defer?  You only receive the pebble ready event once so, if you don't execute the lines that follow this point they would never be executed.

Reply:
Updated, thanks @quiet-ranger